File: Linked List/linked_list_reverse.cpp
// Reverse a linkedlist iteratively
// Program Author : Abhisek Kumar Gupta 
#include<bits/stdc++.h>
using namespace std;
class node{
    public:
        int data;
        node* next;

        node(int d){
            data = d;
            next = NULL;
        }
};
void insert_at_tail(node *&head, int data){
    if(head == NULL){
        head = new node(data);
        return;
    }
    node *n = new node(data);
    node * temp = head;
    while(temp -> next != NULL){
        temp = temp->next;
    }
    temp->next = n;
}
void print_linked_list(node *head){
    while(head != NULL){
        cout << head->data << "->";
        head = head->next;
    }
}
void makeLinkedList(node *&head){
    int data;
    cin >> data;
    while(data != -1){
        insert_at_tail(head, data);
        cin >> data;
    }
}
void reverse_linked_list(node *&head){
    node *current = head;
    node *next = NULL;
    node *prev = NULL;
    while(current != NULL){
        next = current->next;
        current->next = prev;
        prev = current;
        current = next;
    }
    head = prev;
}
int main(){
    node *head = NULL;
    makeLinkedList(head);
    print_linked_list(head);
    cout << endl;
    reverse_linked_list(head);
    print_linked_list(head); 
    return 0;
}

File: Linked List/linked_list_reverse.js
//Problem statement-Given the head of a singly linked list, reverse the list, and return the reversed list.

// Iteratively 
* @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
  // Handling the edge cases if head node is null or when single node is there 
        if (head === null || head.next === null) {
            return head;
        }
  // Using three pointer approach 
        let prev = null;
        let curr = head;
        let fwd = null;
        while (curr !== null) {
          // Updatng the value of each of the pointers 
            fwd = curr.next;
            curr.next = prev;
            prev = curr;
            curr = fwd;
        }
  // Returning the head of the reversed linked list 
        return prev;    
};
// Time complexity - O(N)
// Space complexity - O(1)

// Recursively 
var reverseList = function(head) {
  // Base case- If the head is null or single node is there
  if (head === null || head.next === null) {
    return head;
  }

  // Recursively reverse the remaining list starting from the next node 
  const reversedList = reverseList(head.next);

  // Reversing the links between the current and next node 
  head.next.next = head;
  head.next = null;

  // Returning the head of the reversed linked list 
  return reversedList; 
};
// Time complexity - O(N)
// Space complexity - O(N)
